amm/contracts/stable_pool/lib.rs

```diff
@@ -0,0 +1,1033 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+mod token_rate;
+
+#[ink::contract]
+pub mod stable_pool {
+    use crate::token_rate::TokenRate;
+    use amm_helpers::{
+        constants::stable_pool::{
+            ADMIN_FEE_BPS, MAX_AMP, MIN_AMP, RATE_PRECISION, TOKEN_TARGET_DECIMALS, TRADE_FEE_BPS,
+        },
+        ensure,
+        stable_swap_math::{self as math, fees::Fees},
+    };
+    use ink::contract_ref;
+    use ink::prelude::{
+        string::{String, ToString},
+        {vec, vec::Vec},
+    };
+    use psp22::{PSP22Data, PSP22Error, PSP22Event, PSP22Metadata, PSP22};
+    use traits::{MathError, StablePool, StablePoolError, StablePoolView};
+
+    #[ink(event)]
+    pub struct AddLiquidity {
+        #[ink(topic)]
+        pub provider: AccountId,
+        pub token_amounts: Vec<u128>,
+        pub shares: u128,
+        #[ink(topic)]
+        pub to: AccountId,
+    }
+
+    #[ink(event)]
+    pub struct RemoveLiquidity {
+        #[ink(topic)]
+        pub provider: AccountId,
+        pub token_amounts: Vec<u128>,
+        pub shares: u128,
+        #[ink(topic)]
+        pub to: AccountId,
+    }
+
+    #[ink(event)]
+    pub struct Swap {
+        #[ink(topic)]
+        pub sender: AccountId,
+        pub token_in: AccountId,
+        pub amount_in: u128,
+        pub token_out: AccountId,
+        pub amount_out: u128,
+        #[ink(topic)]
+        pub to: AccountId,
+    }
```

Comment:
Yeah, the rustfmt.toml is very basic here, also `cargo fmt` is used instead of `cargo +nightly fmt` in the makefile. I can change that, but in that case I need a pointer where to copy the config from.